operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1797 @@
+					err := ipam.handleUpsertService(svc)
+					if err != nil {
+						// Let the resource retry after the initialization
+						return err
+					}
+
+					return nil
+				},
+				func(k resource.Key, svc *slim_core_v1.Service) error {
+					ipam.logger.Error("Got a Service deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) poolOnUpsert(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	// Deep copy so we get a version we are allowed to update
+	pool = pool.DeepCopy()
+
+	var err error
+	if _, exists := ipam.poolStore.GetByName(pool.GetName()); exists {
+		ipam.logger.Debugf("Updating pool '%s'", pool.GetName())
+		err = ipam.handlePoolModified(pool)
+		if err != nil {
+			ipam.logger.WithError(err).Error("Error in handlePoolModified()")
+		}
+	} else {
+		ipam.logger.Debugf("Adding pool '%s'", pool.GetName())
+		err = ipam.handleNewPool(pool)
+		if err != nil {
+			ipam.logger.WithError(err).Error("Error in handleNewPool()")
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	err = ipam.settleConflicts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in settleConflicts()")
+		return fmt.Errorf("settleConflicts: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) poolOnDelete(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	err := ipam.handlePoolDeleted(pool)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handlePoolDeleted()")
+		return fmt.Errorf("handlePoolDeleted: %w", err)
+	}
+
+	err = ipam.settleConflicts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in settleConflicts()")
+		return fmt.Errorf("settleConflicts: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) svcOnUpsert(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
+
+	ipam.logger.Debugf("Added/updated service '%s/%s'", svc.GetNamespace(), svc.GetName())
+
+	err := ipam.handleUpsertService(svc)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handleUpsertService()")
+		return fmt.Errorf("handleUpsertService: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) svcOnDelete(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
```

Comment:
From what I can see we are not modifying the `svc` object so we don't need to perform DeepCopy.

Reply:
We are already deep copying the status so this full copy is indeed not necessary, removed!